### DISCUSSION TITLE
SAF (try SAF clone repo to Termux dir)

### DISCUSSION THREAD

**Discussion by Willie169** [catpuppyapp/PuppyGit#1] [General]
I think Termux $PREFIX and some other stuff can be accessed through the Storage Access Framework. The code might look like this:
```
import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.DocumentsContract

class MainActivity : Activity() {

    private val REQUEST_CODE_OPEN_DIRECTORY = 1001
    private var fullPath: String? = null

    // 选择一个目录，授予读写权限以供将来访问
    fun openDirectoryPicker() {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE).apply {
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
        }
        startActivityForResult(intent, REQUEST_CODE_OPEN_DIRECTORY)
    }

    // 接收意图的结果
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE_OPEN_DIRECTORY && resultCode == Activity.RESULT_OK) {
            val treeUri: Uri? = data?.data
            if (treeUri != null) {
                contentResolver.takePersistableUriPermission(
                    treeUri,
                    Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION
                )

                // 将 document id 或 uri 存储在字符串中
                fullPath = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    DocumentsContract.getTreeDocumentId(treeUri)
                } else {
                    treeUri.toString()
                }
            }
        }
    }
}
```
Maybe I'm wrong cuz I haven't tested.

**Comment by Bandeapart1964** [comment]
ok, I tested, is not worked, because puppygit use c clone repo, and SAF can't work in c code. (even in java , SAF use difference api to File though)

I'll try make puppy git can run git command by Termux. if possible 😉

**Comment by Bandeapart1964** [comment]
Can you tell me why you want to clone the repository to the termux's Home folder?

**Reply by Willie169** [reply]
Cuz I usually work in Termux or QEMU VMs in it with help of CMake, npm, gradle, etc., which might not be available or might be less convenient to use outside of it due to:

<table>
<tbody><tr>
<th>Storage type
</th>
<th>chmod/chown support
</th>
<th>Special files support
</th>
<th>Executables support
</th>
<th>Access mode
</th></tr>
<tr>
<td>Internal ($HOME/$PREFIX)
</td>
<td>yes
</td>
<td>yes
</td>
<td>yes
</td>
<td>RW (app dir)
</td></tr>
<tr>
<td>Shared storage
</td>
<td>no
</td>
<td>no
</td>
<td>no
</td>
<td>RW
</td></tr>
<tr>
<td>External storage
</td>
<td>no
</td>
<td>no
</td>
<td>no
</td>
<td>RO / RW (app dir)
</td></tr></tbody></table>

However, I think this feature may not be that important cuz it's also easy to define some aliases in `.bashrc` to make `git` process easier.

**Reply by Bandeapart1964** [reply]
ok, if you need this, even puppy git support run command by termux, still no help to you.

**Reply by Willie169** [reply]
I originally thought that might help those uncomfortable with CLI to use git, but yep, that implies they're not comfortable with Termux either, lol. And I am comfortable with whatever stuff that can be displayed on a phone screen and controlled by a thumb on a crowded metro, where I always code on. Anyway, thanks for your patience.

**Reply by Bandeapart1964** [reply]
> I originally thought that might help those uncomfortable with CLI to use git, but yep, that implies they're not comfortable with Termux either, lol.

you are right. 😂
